### PR TITLE
Include/exclude parameters to V2 validation

### PR DESCRIPTION
**Pull request by nrydanov**

### CODE REVIEW

File: src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/SwitchMapper.java [hunk 1]
```diff
@@ -417,11 +419,46 @@ public RulesValidationDto toRulesValidationDtoV1(RulesValidationEntryV2 data) {
 
     public abstract GroupInfoDtoV2 toGroupInfoDtoV2(GroupInfoEntryV2 data);
 
+    private IncludeFilter toIncludeFilter(String value) {
+        switch (value) {
+            case("meters"):
```
Comment (niksv) [inline src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/SwitchMapper.java:424]: please use toLowerCase

File: src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/SwitchMapper.java [hunk 2]
```diff
@@ -417,11 +419,46 @@ public RulesValidationDto toRulesValidationDtoV1(RulesValidationEntryV2 data) {
...
+            case("rules"):
+                return IncludeFilter.RULES;
+            default:
+                throw new IllegalArgumentException(String.format("Unexpected include filter (%s)", value));
+        }
+    }
+
+    private ExcludeFilter toExcludeFilter(String value) {
+        switch (value) {
+            case("flow_info"):
```
Comment (niksv) [inline src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/SwitchMapper.java:439]: same

File: src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/service/impl/SwitchServiceImpl.java [hunk 3]
```diff
@@ -274,11 +278,30 @@ public CompletableFuture<SwitchValidationResult> validateSwitch(SwitchId switchI
...
+    public CompletableFuture<SwitchValidationResultV2> validateSwitch(SwitchId switchId,
+                                                                      String includeString,
+                                                                      String excludeString) {
         logger.info("Validate api V2 request for switch {}", switchId);
 
+        List<IncludeFilter> includeFilters = new LinkedList<>();
+        if (includeString != null) {
+            includeFilters = switchMapper
+                    .toIncludeFilters(Arrays.stream(includeString.split("\\|"))
+                            .collect(Collectors.toList()));
```
Comment (niksv) [inline src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/service/impl/SwitchServiceImpl.java:290]: please use set

File: src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java [hunk 4]
```diff
@@ -218,15 +249,20 @@ public ValidateLogicalPortsResultV2 validateLogicalPorts(SwitchId switchId, List
...
-                                                 List<MeterSpeakerData> expectedMeterSpeakerData) {
+                                                 List<MeterSpeakerData> expectedMeterSpeakerData,
+                                                 boolean excludeFlowInfo) {
         log.debug("Validating meters on a switch {}", switchId);
 
         Switch sw = switchRepository.findById(switchId)
                 .orElseThrow(() -> new SwitchNotFoundException(switchId));
         boolean isESwitch = Switch.isNoviflowESwitch(sw.getOfDescriptionManufacturer(), sw.getOfDescriptionHardware());
+        Collection<Flow> flows = flowRepository.findAll();
+        Collection<YFlow> yFlows = yFlowRepository.findAll();
```
Comment (niksv) [inline src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java:260]: we can request only flow IDs here and use set of strings to store it

File: src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java [hunk 5]
```diff
@@ -392,33 +425,60 @@ private void processLogicalPortValidation(Map<Integer, LogicalPortInfoEntryV2> a
...
+    }
+
+    private Set<GroupInfoEntryV2> convertGroups(List<GroupSpeakerData> groupEntries, Collection<Flow> flows,
+                                                Collection<YFlow> yFlows, boolean excludeFlowInfo) {
         return groupEntries.stream()
-                .map(GroupEntryConverter.INSTANCE::toGroupEntry)
+                .map(data -> {
+                    GroupInfoEntryV2 groupEntry = GroupEntryConverter.INSTANCE.toGroupEntry(data);
+                    if (!excludeFlowInfo) {
+                        mirrorGroupRepository.findByGroupId(data.getGroupId()).ifPresent(mirrorGroup -> {
```
Comment (niksv) [inline src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java:438]: use findBySwitchIdAndGroupId

File: src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java [hunk 6]
```diff
@@ -392,33 +425,60 @@ private void processLogicalPortValidation(Map<Integer, LogicalPortInfoEntryV2> a
...
+                            String id = mirrorGroup.getFlowId();
+                            if (isFlowId(id, flows)) {
+                                groupEntry.setFlowId(id);
+                            } else {
+                                yFlowRepository.findYFlowId(id).ifPresent(groupEntry::setYFlowId);
```
Comment (niksv) [inline src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java:443]: please use set if yflow IDs

File: src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java [hunk 7]
```diff
@@ -392,33 +425,60 @@ private void processLogicalPortValidation(Map<Integer, LogicalPortInfoEntryV2> a
...
-                flowPath.ifPresent(path -> meterInfoEntry.setCookie(path.getCookie().getValue()));
-
-                meterInfoEntry.setYFlowId("TBD");
-                //TODO: set yFlowId from repo
+            if (flowMeter.isPresent() && !excludeFlowInfo) {
+                String id = flowMeter.get().getFlowId();
+                if (isFlowId(id, flows)) {
+                    meterInfoEntry.setFlowId(id);
+                } else {
+                    yFlowRepository.findYFlowId(id).ifPresent(meterInfoEntry::setYFlowId);
```
Comment (niksv) [inline src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java:467]: same

File: src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java [hunk 8]
```diff
@@ -428,6 +488,34 @@ private Map<RuleKey, RuleInfoEntryV2> convertRules(List<FlowSpeakerData> flowSpe
...
+
+            if ((type == SERVICE_OR_FLOW_SEGMENT || type == SERVER_42_FLOW_RTT_INGRESS)
+                    && segmentCookie.getDirection() != FlowPathDirection.UNDEFINED) {
+                FlowSegmentCookie pureCookie = FlowSegmentCookie.builder()
+                        .direction(segmentCookie.getDirection())
+                        .flowEffectiveId(segmentCookie.getFlowEffectiveId())
+                        .type(SERVICE_OR_FLOW_SEGMENT)
+                        .build();
+
+                if (!excludeFlowInfo) {
```
Comment (niksv) [inline src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java:504]: I think this if should be first

File: src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java [hunk 9]
```diff
@@ -428,6 +488,34 @@ private Map<RuleKey, RuleInfoEntryV2> convertRules(List<FlowSpeakerData> flowSpe
...
+                        ruleInfo.setFlowPathId(flowPath.getPathId().getId());
+
+                        Flow flow = flowPath.getFlow();
+                        String yFlowId = flow.getYFlowId();
+
+                        if (yFlowId != null) {
+                            ruleInfo.setYFlowId(yFlowId);
+                        } else {
+                            ruleInfo.setFlowId(flow.getFlowId());
+                        }
```
Comment (niksv) [inline src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java:515]: cookie can has flow and y flow at the same time

File: src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/SwitchMapper.java (hunk 1 above)
Comment (nrydanov) [reply src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/SwitchMapper.java:424]: Ok

File: src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/service/impl/SwitchServiceImpl.java (hunk 3 above)
Comment (nrydanov) [reply src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/service/impl/SwitchServiceImpl.java:290]: Yep, agree with you

File: src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java (hunk 8 above)
Comment (nrydanov) [reply src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java:504]: yep

**Review by vshakirova — APPROVED** [review]

**Pull request merged**